### PR TITLE
part1 implementation done

### PR DESCRIPTION
**Pull request by kramosik**

### CODE REVIEW

File: AttackItem.h [hunk 1]
```diff
@@ -1 +1,9 @@
-#include "Item.h"
\ No newline at end of file
+#pragma once
+
+#include "Item.h"
+
+struct AttackItem : Item
+{
+    AttackItem() : Item("attack item", 2) {}
```
Comment (matkatmusic) [inline AttackItem.h]: This should have a boost of 10, not 2. 

- [x] adjust boost (2nd param) and revise usage in Utility.cpp

File: Character.cpp [hunk 2]
```diff
@@ -86,32 +89,36 @@ int Character::takeDamage(int damage)
...
       */
-        assert(false);
-        std::cout << getName() << " defeated " << other.getName() << " and leveled up!" << std::endl;        
+        hitPoints = static_cast<int>(*initialHitPoints * 1.1);
+        armor = static_cast<int>(*initialArmorLevel * 1.1);
+        attackDamage = static_cast<int>(*initialAttackDamage * 1.1);
+
+        initialHitPoints = std::make_unique<int>(hitPoints);
+        initialArmorLevel = std::make_unique<int>(armor);
+        initialAttackDamage = std::make_unique<int>(attackDamage);
```
Comment (matkatmusic) [inline Character.cpp]: You don't need to make new instances of these.  Dereference the pointer to update
- [x] FIXME

File: Character.h [hunk 3]
```diff
@@ -9,66 +9,70 @@
...
+    {
+        return defensiveItems;
+    }
+
+    void boostArmor(int amount)
     {
         armor += amount;
-        std::cout << getName() << "'s armor level has been boosted to " << armor << std::endl;
+        std::cout << getName() << "'s armor level has been boosted to " << armor
+                  << std::endl;
```
Comment (matkatmusic) [inline Character.h:47]: Does your clang-format do this automatically?

File: Character.h [hunk 4]
```diff
@@ -9,66 +9,70 @@
...
-    void boostArmor( int amount )
+
+    const std::vector<std::unique_ptr<Item>>& getHelpfulItems() const
+    {
+        return helpfulItems;
+    }
+    const std::vector<std::unique_ptr<Item>>& getDefensiveItems() const
+    {
+        return defensiveItems;
+    }
```
Comment (matkatmusic) [inline Character.h:41]: Did your clang-format change this?

File: Character.h [hunk 5]
```diff
@@ -9,66 +9,70 @@
 
 struct Character
 {
-    Character(int hp, int armor_, int attackDamage_ );
-    virtual ~Character() { }
-    
+    Character(int hp, int armor_, int attackDamage_);
+    virtual ~Character() = default;
```
Comment (matkatmusic) [inline Character.h:13]: Did clang-format do this change or did you watch the Chapter 5 vids that cover making special member functions `= default;` ?

File: Paladin.cpp [hunk 6]
```diff
@@ -1,7 +1,23 @@
 #include "Paladin.h"
+#include "Utility.h"
+#include <random>
```
Comment (matkatmusic) [inline Paladin.cpp]: Don't include what you don't use!

File: Utility.cpp [hunk 7]
```diff
@@ -1,108 +1,141 @@
...
+}
+
+std::vector<std::unique_ptr<Item>> makeRandomDefensiveItems()
+{
+    std::random_device rd;
+    std::mt19937 gen(rd());
+    std::uniform_int_distribution<int> dist(MIN_RANDOM_ITEM, MAX_RANDOM_ITEM);
+
+    return makeDefensiveItems(dist(gen));
+}
```
Comment (matkatmusic) [inline Utility.cpp]: This seems overly complicated.  Why not simply call `make_____Items(rand())` in the derived Character constructors?

File: Utility.cpp [hunk 8]
```diff
@@ -1,108 +1,141 @@
...
+const int MIN_RANDOM_ITEM = 1;
+const int MAX_RANDOM_ITEM = 5;
 
 std::vector<std::unique_ptr<Item>> makeHelpfulItems(int num)
 {
     std::vector<std::unique_ptr<Item>> items;
-    
-    while( num-- >= 0 )
+
+    while (num-- > 0)
```
Comment (matkatmusic) [inline Utility.cpp:16]: Is this a bug fix?   Was the prior `while()` adding one extra element?  if so, nice catch.

File: Utility.cpp [hunk 9]
```diff
@@ -1,108 +1,141 @@
...
-        //check Character.h for available member functions you can use.
+        // DragonSlayers get a 10x boost when attacking dragons, from their
+        // attack item. so their attack item should boost their attack damage by
+        // a factor of 10 this means you need to GET the attack damage, multiply
+        // it by the item's boost, and BOOST the attackDamage with that
+        // multiplied value. check Character.h for available member functions
+        // you can use.
+        auto atkDamage = ch->getAttackDamage();
+        ch->boostAttackDamage(
+            static_cast<int>(atkDamage * item->getBoost() * 10));
```
Comment (matkatmusic) [inline Utility.cpp]: `item->getBoost()` should return 10. You shouldn't need to multiply the boost by anything. 
> their attack item should boost their attack damage by a factor of 10

- [x] Fix this line according to the spec.

**Review by matkatmusic — CHANGES_REQUESTED** [review]

File: Character.h (hunk 5 above)
Comment (kramosik) [reply Character.h:13]: yeah, I have finished most of the videos but the projects are a little behind. I also set up a warning in clang-tidy to catch this.

File: Utility.cpp (hunk 8 above)
Comment (kramosik) [reply Utility.cpp:16]: Yes, I noticed the random generator sometimes gives a character 6 items, but I'm pretty sure the range is set correctly.

File: Character.h (hunk 3 above)
Comment (kramosik) [reply Character.h:47]: yes, I set up a maximum line width

**Review by matkatmusic — APPROVED** [review]

**Pull request merged**